Implement WebView via Playwright

cpiber opened the pull request:
Partially addresses #161

This also contains an implementation of Looper and MessageQueue, since extensions typically put WebView handling on the loop.

All stubs and interfaces are based on recent upstream files as seen on cs.android.com. Thus they do not 100% match the version in the stubs, but since they're backward compatible that should be fine.

This enables the Cubari extension:
![20250607_13h06m05s_grim](https://github.com/user-attachments/assets/2bb5098c-d00b-4d55-9b6c-bc38449f8d2c)

There seems to be a discrepancy somewhere though; an invalid URL should show the message "Unable to parse. Is your query in the format of cubari:<source>/<slug>?", instead we get the following backtrace:

```
12:47:54.572 [DefaultDispatcher-worker-10] ERROR suwayomi.tachidesk.graphql.AsDataFetcherResult -- asDataFetcherResult: failed due to
kotlinx.serialization.json.internal.JsonDecodingException: Cannot read Json element because of unexpected end of the input at path: $
JSON input: 
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
        at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
        at kotlinx.serialization.json.internal.AbstractJsonLexer.fail(AbstractJsonLexer.kt:587)
        at kotlinx.serialization.json.internal.AbstractJsonLexer.fail$default(AbstractJsonLexer.kt:585)
        at kotlinx.serialization.json.internal.JsonTreeReader.read(JsonTreeReader.kt:107)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeJsonElement(StreamingJsonDecoder.kt:50)
        at kotlinx.serialization.json.JsonElementSerializer.deserialize(JsonElementSerializers.kt:50)
        at kotlinx.serialization.json.JsonElementSerializer.deserialize(JsonElementSerializers.kt:27)
        at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:69)
        at kotlinx.serialization.json.Json.decodeFromString(Json.kt:149)
        at kotlinx.serialization.json.Json.parseToJsonElement(Json.kt:179)
        at eu.kanade.tachiyomi.extension.all.cubari.Cubari.searchMangaParse(Unknown Source)
        at eu.kanade.tachiyomi.extension.all.cubari.Cubari.access$searchMangaParse(Unknown Source)
        at eu.kanade.tachiyomi.extension.all.cubari.Cubari$fetchSearchManga$2.invoke(Unknown Source)
        at eu.kanade.tachiyomi.extension.all.cubari.Cubari$fetchSearchManga$2.invoke(Unknown Source)
        at eu.kanade.tachiyomi.extension.all.cubari.Cubari.fetchSearchManga$lambda$15(Unknown Source)
        at eu.kanade.tachiyomi.extension.all.cubari.Cubari.$r8$lambda$PnjAk5VfRyZhxUCoiHrOSDkWIgE(Unknown Source)
        at eu.kanade.tachiyomi.extension.all.cubari.Cubari$$ExternalSyntheticLambda7.call(Unknown Source)
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
        at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onNext(OnSubscribeDoOnEach.java:101)
        at eu.kanade.tachiyomi.network.OkHttpExtensionsKt$asObservable$1$requestArbiter$1.request(OkHttpExtensions.kt:38)
        at rx.Subscriber.setProducer(Subscriber.java:211)
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
        at rx.Subscriber.setProducer(Subscriber.java:205)
        at eu.kanade.tachiyomi.network.OkHttpExtensionsKt.asObservable$lambda$0(OkHttpExtensions.kt:56)
        at rx.Observable.unsafeSubscribe(Observable.java:10327)
        at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41)
        at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30)
        at rx.Observable.unsafeSubscribe(Observable.java:10327)
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
        at rx.Observable.unsafeSubscribe(Observable.java:10327)
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
        at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.Observable.subscribe(Observable.java:10423)
        at rx.Observable.subscribe(Observable.java:10390)
        at suwayomi.tachidesk.manga.impl.util.lang.RxCoroutineBridgeKt.awaitOne(RxCoroutineBridge.kt:28)
        at suwayomi.tachidesk.manga.impl.util.lang.RxCoroutineBridgeKt.awaitSingle(RxCoroutineBridge.kt:22)
        at eu.kanade.tachiyomi.source.CatalogueSource$DefaultImpls.getSearchManga(CatalogueSource.kt:41)
        at eu.kanade.tachiyomi.source.online.HttpSource.getSearchManga(HttpSource.kt:28)
        at suwayomi.tachidesk.graphql.mutations.SourceMutation$fetchSourceManga$1.invokeSuspend(SourceMutation.kt:128)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
        at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:124)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:89)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:820)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:717)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:704)
Caused by: rx.exceptions.OnErrorThrowable$OnNextValue: OnError while emitting onNext value: okhttp3.Response.class
        at rx.exceptions.OnErrorThrowable.addValueAsLastCause(OnErrorThrowable.java:118)
        at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:73)
        ... 33 common frames omitted
```
The extension seems to assume errors are translated into empty lists, and this appears to work on Mihon
https://github.com/keiyoushi/extensions-source/blob/bdb776bfc9844a0954dbb48a6929f3e605217885/src/all/cubari/src/eu/kanade/tachiyomi/extension/all/cubari/Cubari.kt#L245-L256

Most other extensions that use WebView are **not yet** enabled by this. This implementation stubs most settings, some of which are used. More importantly, events like `shouldInterceptRequest` are not yet implemented. I would appreciate a review of the existing code while I see into implementing more parts.

cpiber committed:
Implement Android's Looper

Looper handles thread messaging. This is used by extensions when they
want to enqueue actions e.g. for sleeping while WebView does someting

cpiber committed:
Stub WebView

cpiber committed:
Continue stubbing ViewGroup for WebView

cpiber committed:
Implement WebView via Playwright

cpiber committed:
Lint

cpiber committed:
Implement request interception

Supports Yidan

cpiber committed:
Support WebChromeClient

For Bokugen

cpiber committed:
Fix onPageStarted

cpiber committed:
Make Playwright configurable

cpiber committed:
Subscribe to config changes

cpiber committed:
Fix exposing of functions

cpiber committed:
Support data urls

cpiber committed:
Looper: Fix infinite sleep

cpiber committed:
Looper: Avoid killing the loop on exception

Just log it and continue

cpiber committed:
Pump playwright's message queue periodically

https://playwright.dev/java/docs/multithreading#pagewaitfortimeout-vs-threadsleep

Syer10 commented on AndroidCompat/src/main/java/android/webkit/PlaywrightWebSettings.java:
```diff
@@ -0,0 +1,534 @@
+        throw new RuntimeException("Stub!");
+    }
+
+    public void setUserAgentString(@Nullable String ua) {
```
Do we call this, or does the extension? If its the extension we need to change the default to NetworkHelper.userAgent, either by moving the userAgent MutableStateFlow to AndroidCompat or by setting a system property.

cpiber replied:
The extension does.

Good to know, will change that. I think a setter is better, and it's also how I solved configuring the browser.

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/graphql/types/SettingsType.kt:
```diff
@@ -208,6 +217,10 @@ class SettingsType(
+    // playwright webview
+    override val playwrightBrowser: String?,
+    override val playwrightWsEndpoint: String?,
+    override val playwrightSandbox: Boolean?,
```
```suggestion
    override val playwrightBrowser: String,
    override val playwrightWsEndpoint: String,
    override val playwrightSandbox: Boolean,
```

Syer10 commented on server/src/main/resources/server-reference.conf:
```diff
@@ -32,6 +32,11 @@ server.extensionRepos = [
 ]
 
+# playwright webview
+server.playwrightBrowser = "chromium"
```
Add a comment with possible browser values

Syer10 commented on AndroidCompat/build.gradle.kts:
```diff
@@ -21,6 +21,8 @@ dependencies {
     implementation(libs.bundles.shared)
     testImplementation(libs.bundles.sharedTest)
 
+    implementation("com.microsoft.playwright:playwright:1.51.0")
```
This should be moved to gradle/libs.versions.toml

cpiber replied:
Ups, yeah, did that for testing, forgot to move.

Syer10 reviewed (commented):
This is looking great! While KCEF would be preferred, Playwright is a good alternative, we have used it previously as an attempt to bypass Cloudflare, you should checkout the previous PRs before it was eventually removed. https://github.com/Suwayomi/Suwayomi-Server/pulls?q=is%3Apr+playwright+is%3Aclosed

One bad thing about Playwright is that it downloads Chromium/the browser all on its own, we previously tried adding it to the download package but that just sorta worked, it wasn't worth it, we need a pre-download initializer at least so that Suwayomi doesn't just stop in the middle of a request while it downloads.
Docker also has an issue where it will download to a unmounted folder, so it redownloads the browser on every re-create of the container, changing the browser download to somewhere in the root directory would be preferred.

For cloudflare issues we can re-add the Cloudsflare bypass scripts we tested with Playwright previously to help it bypass Cloudflare(see https://github.com/Suwayomi/Suwayomi-Server/pull/456), it just wasn't the optimial choice due to the above issues so it was removed.

JxBrowser is not possible since we won't distribute pirated or paid versions of any library in the app.

In general well done! I'm looking forward to this being improved so we can merge it.

cpiber committed:
Update server/src/main/kotlin/suwayomi/tachidesk/graphql/types/SettingsType.kt

Co-authored-by: Mitchell Syer <Syer10@users.noreply.github.com>

cpiber committed:
Stub a KCef WebViewProvider

cpiber committed:
Initial Kcef Webview implementation

Still buggy, on the second call it just seems to fall over

cpiber committed:
Format, restructure to create browser on load

This is much more consistent, before we would sometimes see errors from
about:blank, which block the actual page

cpiber committed:
Implement some small useful properties

cpiber committed:
Move inline objects to class

cpiber committed:
Handle requests in Kcef

cpiber committed:
Move Playwright implementation

cpiber committed:
Document Playwright settings, fix deprecated warnings

cpiber committed:
Inject default user agent from NetworkHelper

cpiber committed:
Move playwright to libs.versions.toml

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:
```diff
@@ -187,7 +227,10 @@ fun applicationSetup() {
-    Injekt.get<NetworkHelper>()
+    Injekt.get<NetworkHelper>().userAgentFlow.onEach {
+        SystemProperties.set("http.agent", it)
+    }
```
Need to launch the listener
```suggestion
    }.launchIn(GlobalScope)
```

cpiber replied:
Thanks, I have never worked with StateFlow

Syer10 commented on AndroidCompat/src/main/java/xyz/nulldev/androidcompat/AndroidCompatInitializer.kt:
```diff
@@ -17,6 +19,8 @@ class AndroidCompatInitializer {
             SystemConfigModule.register(GlobalConfigManager.config),
         )
 
+        WebView.setProviderFactory({ view: WebView -> KcefWebViewProvider(view) })
```
I would move this to ServerSetup and make it a user configurable setting, so that a user can choose Playwright or KCEF

cpiber replied:
Done. I had to leave it there because when trying to use those methods in ServerSetup, I got some weird linking errors.

Do we have to update android-lib for this? It seemed like it was complaining about differences in the WebView stub and the new compat implementation

```
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:117:33 Cannot access 'Callback' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:117:33 Cannot access 'AccessibilityEventSource' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:117:33 Cannot access 'ViewParent' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:117:33 Cannot access 'ViewManager' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:117:33 Cannot access 'OnGlobalFocusChangeListener' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:118:38 Cannot access 'Callback' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:118:38 Cannot access 'AccessibilityEventSource' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:118:38 Cannot access 'ViewParent' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:118:38 Cannot access 'ViewManager' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
e: file:///.../Suwayomi/Suwayomi-Server/server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt:118:38 Cannot access 'OnGlobalFocusChangeListener' which is a supertype of 'WebView'. Check your module classpath for missing or conflicting dependencies.
```

Syer10 replied:
Looks like it, I will update the android library after I merge this. This solution will work though so its fine.

cpiber committed:
Lint

Syer10 reviewed (commented):
Its looking good! A few comments, I like that you added both implementations, we can now make it user configurable with KCEF as a default

cpiber committed:
Fix missing imports after lint

cpiber committed:
Update server/src/main/kotlin/suwayomi/tachidesk/server/ServerSetup.kt

Co-authored-by: Mitchell Syer <Syer10@users.noreply.github.com>

cpiber committed:
Fix default user agent set/get

Use System.getProperty instead of SystemProperties.get

cpiber committed:
Configurable WebView provider implementation

cpiber committed:
Simplify Playwright settings init

Syer10 reviewed (approved):
Looks good! I will merge it after I get the docker env variables and launcher settings added. I will also verify that KCEF works fine in the docker container.

Syer10 committed:
Minor cleanup and improvements

cpiber committed:
Remove playwright WebView impl

cpiber committed:
Document WebView for Linux
